keps/sig-storage/2639-secret-configmap-protection/README.md

```diff
@@ -0,0 +1,652 @@
+    - provisioner secret
+    - controller publish secret
+    - node stage secret
+    - node publish secret
+    - controller expand secret
+  - non-CSI:
+    - dependent on each storage driver and will be deprecated soon (Out of scope)
+- [From VolumeSnapshot](https://kubernetes-csi.github.io/docs/secrets-and-credentials-volume-snapshot-class.html):
+  - snapshotter secret
+
+ConfigMaps can be used by below ways:
+- From Pod:
+  - [Mounted as files](https://kubernetes.io/docs/concepts/configuration/configmap/#using-configmaps-as-files-from-a-pod)
+  - [Exposed as environment variables](https://kubernetes.io/docs/concepts/configuration/configmap/#configmaps-and-pods)
+- From Deployment:
+  - Specified through [DeploymentSpec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.22/#deploymentspec-v1-apps) -> [PodTemplateSpec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.22/#podtemplatespec-v1-core)
+- From StatefulSet:
+  - Specified through [StatefulSetSpec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.22/#statefulsetspec-v1-apps) -> [PodTemplateSpec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.22/#podtemplatespec-v1-core)
+- From DaemonSet:
+  - Specified through [DaemonSetSpec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.22/#daemonsetspec-v1-apps) -> [PodTemplateSpec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.22/#podtemplatespec-v1-core)
+- From CronJob:
+  - Specified through [JobTemplateSpec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.22/#cronjob-v1-batch) -> [JobSpec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.22/#jobspec-v1-batch) -> [PodTemplateSpec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.22/#podtemplatespec-v1-core)
+
+### Goals
+
+- Protect Secrets/ConfigMaps from being deleted while they are in use.
+
+### Non-Goals
+
+- Protect important Secrets/ConfigMaps that aren't in use from being deleted
+- Protect Secrets/ConfigMaps from being __updated__ while they are in use (Immutable Secrets/ConfigMaps will solve it)
+- Protect resources other than Secrets/ConfigMaps from being deleted.
+- Provide a generic mechanism that prevents a resource from being deleted when it shouldn't be deleted (Scope of [KEP-2839](https://github.com/kubernetes/enhancements/pull/2840))
+
+## Proposal
+
+New controllers to protect Secrets/ConfigMaps are introduced.
+
+### User Stories
+
+#### Story 1
+
+A user creates Secrets/ConfigMaps and a pod using them. Then, the user mistakenly delete them while the pod is using them.
+They are protected until the pod using them is deleted.
+
+#### Story 2
+
+A user creates a volume that uses a certain secret in the same namespace. Then, the user delete the namespace.
+The secret is protected until the volume using the secret is deleted and the deletion of the volume succeeds.
```

Comment:
Is story 2 trying to protect the cluster from actions taken by the user? if so, story 3 seems to allow bypassing/undoing the solution for story 2.

It seems like placing the resources required to administrate a volume backing a PVC in a namespace other than the PVC would resolve this issue and be simpler.